### PR TITLE
mgr: fixing dashboard configuration handling

### PR DESCRIPTION
**Pull request by rkachach**
previously, the dashboard parameters supported by Rook were stored in the daemon configuration section (mgr.X, for example). This differs from Cephadm-based deployments, where all configurations are stored in the global mgr configuration section. This variance could result in configuration mismatches between the active and standby dashboards. Furthermore, all Ceph dashboard documentation exclusively points to the global mgr configuration section and makes no use of individual daemons sections.

Fixes: https://github.com/rook/rook/issues/13577

**Upgrade test**

Starting from previous vesion (`image: rook/ceph:master`) and using the following configuration:

```
  mgr:
    count: 2
    allowMultiplePerNode: true
  dashboard:
    enabled: true
    port: 7777
    ssl: false
    prometheusEndpoint: "http://192.168.39.54:30900"
    prometheusEndpointSSLVerify: false
```

Let's get the status of the config once the cluster is up&running:

```
bash-4.4$ ceph config get mgr.a
WHO     MASK  LEVEL     OPTION                                   VALUE                       RO
mgr.a         advanced  mgr/dashboard/PROMETHEUS_API_HOST        http://192.168.39.54:30900  * 
mgr.a         advanced  mgr/dashboard/PROMETHEUS_API_SSL_VERIFY  false                       * 
mgr.a         advanced  mgr/dashboard/server_port                7777                        * 
mgr.a         advanced  mgr/dashboard/ssl                        false                       * 
mgr           advanced  mgr/orchestrator/orchestrator            rook                          
mgr           advanced  mgr/prometheus/rbd_stats_pools                                       * 

bash-4.4$ ceph config get mgr.b
WHO     MASK  LEVEL     OPTION                                   VALUE                       RO
mgr.b         advanced  mgr/dashboard/PROMETHEUS_API_HOST        http://192.168.39.54:30900  * 
mgr.b         advanced  mgr/dashboard/PROMETHEUS_API_SSL_VERIFY  false                       * 
mgr.b         advanced  mgr/dashboard/server_port                7777                        * 
mgr.b         advanced  mgr/dashboard/ssl                        false                       * 
mgr           advanced  mgr/orchestrator/orchestrator            rook                          
mgr           advanced  mgr/prometheus/rbd_stats_pools                                       * 

bash-4.4$ ceph config get mgr  
WHO     MASK  LEVEL     OPTION                             VALUE                        RO
mgr           advanced  mgr/dashboard/PROMETHEUS_API_HOST  http://192.168.39.215:30900  * 
mgr           advanced  mgr/orchestrator/orchestrator      rook                           
mgr           advanced  mgr/prometheus/rbd_stats_pools                                  * 
bash-4.4$ 

```

Mgr daemons configuration after the upgrade (no specific per-daemon configuration anymore). The configuration related to ssl now is part of the main mgr section.

```
bash-4.4$ ceph config get mgr.a
WHO     MASK  LEVEL     OPTION                                   VALUE                       RO
mgr           advanced  mgr/dashboard/PROMETHEUS_API_HOST        http://192.168.39.54:30900  * 
mgr           advanced  mgr/dashboard/PROMETHEUS_API_SSL_VERIFY  false                       * 
mgr           advanced  mgr/dashboard/server_port                7777                        * 
mgr           advanced  mgr/dashboard/ssl                        false                       * 
mgr           advanced  mgr/orchestrator/orchestrator            rook                          
mgr           advanced  mgr/prometheus/rbd_stats_pools                                       * 

bash-4.4$ 
bash-4.4$ ceph config get mgr.b
WHO     MASK  LEVEL     OPTION                                   VALUE                       RO
mgr           advanced  mgr/dashboard/PROMETHEUS_API_HOST        http://192.168.39.54:30900  * 
mgr           advanced  mgr/dashboard/PROMETHEUS_API_SSL_VERIFY  false                       * 
mgr           advanced  mgr/dashboard/server_port                7777                        * 
mgr           advanced  mgr/dashboard/ssl                        false                       * 
mgr           advanced  mgr/orchestrator/orchestrator            rook                          
mgr           advanced  mgr/prometheus/rbd_stats_pools                                       * 
bash-4.4$ 

bash-4.4$ ceph config get mgr  
WHO     MASK  LEVEL     OPTION                                   VALUE                       RO
mgr           advanced  mgr/dashboard/PROMETHEUS_API_HOST        http://192.168.39.54:30900  * 
mgr           advanced  mgr/dashboard/PROMETHEUS_API_SSL_VERIFY  false                       * 
mgr           advanced  mgr/dashboard/server_port                7777                        * 
mgr           advanced  mgr/dashboard/ssl                        false                       * 
mgr           advanced  mgr/orchestrator/orchestrator            rook                          
mgr           advanced  mgr/prometheus/rbd_stats_pools                                       * 
```

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Review by jmolmo — APPROVED** [review]
Good catch!!! LGTM.
Just review we are not "getting" dashboard options in other operator places.

File: pkg/operator/ceph/cluster/mgr/dashboard.go [hunk 1]
```diff
@@ -110,52 +104,62 @@ func (c *Cluster) configureDashboardModules() error {
...
 }
 
-func (c *Cluster) configureDashboardModuleSettings(daemonID string) (bool, error) {
+func (c *Cluster) configureDashboardModuleSettings() (bool, error) {
+
 	monStore := config.GetMonStore(c.context, c.clusterInfo)
 
-	daemonID = fmt.Sprintf("mgr.%s", daemonID)
+	// Remove any existing per mgr-daemon configuration
+	for _, daemonID := range c.getDaemonIDs() {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mgr/dashboard.go:112]: Before calling all these individual commands, how about if we do a sing `monStore.Get()` to figure out if one of them is set (perhaps the last one in the list), and then only call all these deletes if it is set. Then after the upgrade, future reconciles will not perform so many unnecessary deletes. It's just a small optimization, but one that adds up.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/cluster/mgr/dashboard.go (hunk 1 above)
Comment (rkachach) [reply pkg/operator/ceph/cluster/mgr/dashboard.go:112]: @travisn It's indeed a good idea, unfortunately the get always work even when the entity doesn't exist. For example, let's get the field of the invalid "mgr.z":

```
bash-4.4$ ceph config get mgr.z "mgr/dashboard/server_port"
7777
```
ceph config for whatever reason returns the default values even in this case... sounds like BUG but it could be a "feature" to prevent something from crashing in case the config section doesn't exist.

Comment (travisn) [reply pkg/operator/ceph/cluster/mgr/dashboard.go:112]: Oh right, we can't detect if the setting applies to a specific daemon. 

Then I'd suggest just setting a bool so that it's only done once per time the operator runs. It's fine if the operator restarts and runs the deletes again, but just trying to avoid the deletes with every reconcile.

Comment (rkachach) [reply pkg/operator/ceph/cluster/mgr/dashboard.go]: Done

File: pkg/operator/ceph/cluster/mgr/dashboard.go [hunk 2]
```diff
@@ -110,52 +105,65 @@ func (c *Cluster) configureDashboardModules() error {
...
 }
 
-func (c *Cluster) configureDashboardModuleSettings(daemonID string) (bool, error) {
+func (c *Cluster) configureDashboardModuleSettings() (bool, error) {
+
 	monStore := config.GetMonStore(c.context, c.clusterInfo)
 
-	daemonID = fmt.Sprintf("mgr.%s", daemonID)
+	// Remove any existing per mgr-daemon configuration
+	if removeMgrDaemonConfiguration {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mgr/dashboard.go:113]: How about if we remove these deletions down below so it is after the new settings are configured? Otherwise there may be a brief time where the old settings are deleted and the new settings are not yet applied.

File: pkg/operator/ceph/cluster/mgr/dashboard.go [hunk 3]
```diff
@@ -110,52 +105,65 @@ func (c *Cluster) configureDashboardModules() error {
...
+		for _, daemonID := range c.getDaemonIDs() {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mgr/dashboard.go:114]: How about an overall simple log statement to indicate that the old settings are being removed if they exist. It's helpful for troubleshooting in the operator log what is happening.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/cluster/mgr/dashboard.go (hunk 3 above)
Comment (rkachach) [reply pkg/operator/ceph/cluster/mgr/dashboard.go]: Done :+1:

File: pkg/operator/ceph/cluster/mgr/dashboard.go [hunk 4]
```diff
@@ -110,58 +105,72 @@ func (c *Cluster) configureDashboardModules() error {
...
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/url_prefix")
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/ssl")
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/PROMETHEUS_API_HOST")
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/PROMETHEUS_API_SSL_VERIFY")
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/server_port")
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/ssl_server_port")
+		}
+		removeMgrDaemonConfiguration = false
+	}
+
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/mgr/dashboard.go:173]: At what version of Ceph does this need to be done?

Presumably, the changes in Ceph's code made these redundant at a specific version (or versions), and we should make sure we don't accidentally break older verisons by doing this.

Comment (rkachach) [reply pkg/operator/ceph/cluster/mgr/dashboard.go:173]: @BlaineEXE ceph dashboard (in cephadm based deployments) has always used configuration at mgr level. Not sure why in Rook we were using per-daemon config. This shouldn't be an issue but if for any strange reason there's a mismatch in that conf we could end up in weird issues. 

With this change I'm just removing the usage of per-daemon configs and moving to using the generic mgr section for all the dashboard related conf parameters. 

Does this answer your question?

File: pkg/operator/ceph/cluster/mgr/dashboard.go [hunk 5]
```diff
@@ -110,58 +105,72 @@ func (c *Cluster) configureDashboardModules() error {
...
+	if removeMgrDaemonConfiguration {
+		logger.Info("Removing any previous per mgr-daemon configuration")
+		for _, daemonID := range c.getDaemonIDs() {
+			mgrDaemonID := fmt.Sprintf("%s.%s", config.MgrType, daemonID)
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/url_prefix")
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/ssl")
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/PROMETHEUS_API_HOST")
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/PROMETHEUS_API_SSL_VERIFY")
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/server_port")
+			monStore.Delete(mgrDaemonID, "mgr/dashboard/ssl_server_port")
```
[the remark this answers is not in the document]
Comment (rkachach) [reply pkg/operator/ceph/cluster/mgr/dashboard.go:169]: @BlaineEXE good point, in the first version of the code I was calling this always but then I changed that because of performance concerns. As of the current code, in case we fail to delete some parameter it will be still in use as it has more precedence than the global mgr section, in the worst case this could lead to a mismatch in the configuration. Maybe adding an error log at least could be helpful to detect potential issues.

File: pkg/operator/ceph/cluster/mgr/dashboard.go (hunk 4 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/mgr/dashboard.go:173]: Got it! Yes, I missed the generic vs. specific subtext here.

File: pkg/operator/ceph/cluster/mgr/dashboard.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/mgr/dashboard.go:169]: > it has more precedence than the global mgr section, in the worst case this could lead to a mismatch in the configuration.

I think it may be good to re-reconcile if there are errors given this. Or at minimum, we should not set the "I have already completed this task" boolean that Travis mentioned if there are errors. It may be okay to let this fail with error messages, but we do want to make sure Rook is trying to get to the ideal end state as much as possible so that we don't make it an issue for users where they need to do manual steps.

[the remark this answers is not in the document]
Comment (rkachach) [reply pkg/operator/ceph/cluster/mgr/dashboard.go]: @BlaineEXE I updated the code so the new version checks whether all the delete operations were successful or not before setting the `removeMgrDaemonConfiguration` to false. Just to add that as part of the testing I observed that `ceph config rm` is really checking whether the key exists or not. It only cares about the section it self:

**success even when the key is not valid**
```
bash-4.4$ ceph config rm mgr mgr/invalid_key
bash-4.4$ echo $?
0
```
**success even when the mgr daemon doesn't exist  (mgr.X)**
```
bash-4.4$ ceph config rm mgr.X mgr/dashboard/PROMETHEUS_API_HOST
bash-4.4$ echo $?
0

```
**fails when the section is not valid**
```
bash-4.4$ ceph config rm invalid_section mgr/
Error EINVAL: unrecognized config target 'invalid_section'
```

In summary, in practice the delete will fail in case of timeout or a similar error. The ceph config is not detecting issues with invalid keys. It only detects invalid sections (not subsections such as `mgr.X`).

[the remark this answers is not in the document]
Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/mgr/dashboard.go]: This doesn't take into account what happens if Ceph is having a connection issue during the time where these commands are being sent or if there is a bug in Ceph.

[the remark this answers is not in the document]
Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/mgr/dashboard.go]: Got it! That's good investigation. It's good to make sure that the command is idempotent on Ceph's side. :)

**Review by BlaineEXE — APPROVED** [review]

**Comment by parth-gr** [comment]
```
bash-4.4$ ceph config get mgr  
WHO     MASK  LEVEL     OPTION                                   VALUE                       RO
mgr           advanced  mgr/dashboard/PROMETHEUS_API_HOST        http://192.168.39.54:30900  * 
mgr           advanced  mgr/dashboard/PROMETHEUS_API_SSL_VERIFY  false                       * 
mgr           advanced  mgr/dashboard/server_port                7777                        * 
mgr           advanced  mgr/dashboard/ssl                        false                       * 
mgr           advanced  mgr/orchestrator/orchestrator            rook                          
mgr           advanced  mgr/prometheus/rbd_stats_pools                                       * 
```

This is the active mgr configuration?

File: pkg/operator/ceph/cluster/mgr/dashboard.go [hunk 6]
```diff
@@ -51,7 +51,8 @@ const (
 )
 
 var (
-	dashboardInitWaitTime = 5 * time.Second
+	dashboardInitWaitTime        = 5 * time.Second
+	removeMgrDaemonConfiguration = true
```
Comment (parth-gr) [inline pkg/operator/ceph/cluster/mgr/dashboard.go:55]: do we need this globally?

File: pkg/operator/ceph/cluster/mgr/dashboard.go [hunk 7]
```diff
@@ -110,58 +105,96 @@ func (c *Cluster) configureDashboardModules() error {
...
+	changed, err = monStore.SetIfChanged(config.MgrType, "mgr/dashboard/PROMETHEUS_API_SSL_VERIFY", prometheusEndpointSSLVerify)
 	if err != nil {
 		return false, err
 	}
 	hasChanged = hasChanged || changed
 
 	// server port
 	port := strconv.Itoa(c.dashboardInternalPort())
-	changed, err = monStore.SetIfChanged(daemonID, "mgr/dashboard/server_port", port)
+	changed, err = monStore.SetIfChanged(config.MgrType, "mgr/dashboard/server_port", port)
```
Comment (parth-gr) [inline pkg/operator/ceph/cluster/mgr/dashboard.go:178]: want to add unit test for this

**Comment by rkachach** [comment]
> ```
> bash-4.4$ ceph config get mgr  
> WHO     MASK  LEVEL     OPTION                                   VALUE                       RO
> mgr           advanced  mgr/dashboard/PROMETHEUS_API_HOST        http://192.168.39.54:30900  * 
> mgr           advanced  mgr/dashboard/PROMETHEUS_API_SSL_VERIFY  false                       * 
> mgr           advanced  mgr/dashboard/server_port                7777                        * 
> mgr           advanced  mgr/dashboard/ssl                        false                       * 
> mgr           advanced  mgr/orchestrator/orchestrator            rook                          
> mgr           advanced  mgr/prometheus/rbd_stats_pools                                       * 
> ```
> 
> This is the active mgr configuration?

This the general mgr configuration and it affects to ALL the mgr daemons. Previously we were having the config per-daemon which could lead to some inconsistencies in the conf and is different from what cephadm based deployments (where all the config is at manager level)

File: pkg/operator/ceph/cluster/mgr/dashboard.go [hunk 8]
```diff
@@ -110,58 +105,96 @@ func (c *Cluster) configureDashboardModules() error {
 	return nil
 }
 
-func (c *Cluster) configureDashboardModuleSettings(daemonID string) (bool, error) {
+// Delete the manager per-daemon configuration. Returns true
+// if all the configuration entries have been delete successfully.
+func (c *Cluster) deleteManagerDaemonConfiguration() bool {
```
[the remark this answers is not in the document]
Comment (rkachach) [reply pkg/operator/ceph/cluster/mgr/dashboard.go:110]: No, because we didn't have to. Since now we are moving to a different configuration scheme (at component level instead of daemon level) then we are deleting the previous configuration (per daemon) as it has more priority.

File: pkg/operator/ceph/cluster/mgr/dashboard.go (hunk 6 above)
Comment (rkachach) [reply pkg/operator/ceph/cluster/mgr/dashboard.go:55]: I think the answer is yes because we need this to be stored between reconcile calls.

File: pkg/operator/ceph/cluster/mgr/dashboard.go (hunk 7 above)
Comment (rkachach) [reply pkg/operator/ceph/cluster/mgr/dashboard.go:178]: mmm I think UT only make sense here in this case if we have had mock for the ceph config.

File: pkg/operator/ceph/cluster/mgr/dashboard.go [hunk 9]
```diff
@@ -110,58 +105,96 @@ func (c *Cluster) configureDashboardModules() error {
...
+	}
+
+	success := true
 	monStore := config.GetMonStore(c.context, c.clusterInfo)
+	for _, daemonID := range c.getDaemonIDs() {
+		mgrDaemonID := fmt.Sprintf("%s.%s", config.MgrType, daemonID)
+		for _, key := range mgrKeysToDelete {
+			err := monStore.Delete(mgrDaemonID, key)
+			if err != nil {
+				success = false
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mgr/dashboard.go:128]: Let's add a log statement to show which key failed and what the error was

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/cluster/mgr/dashboard.go (hunk 9 above)
Comment (rkachach) [reply pkg/operator/ceph/cluster/mgr/dashboard.go:128]: @travisn I had that log but then removed it because it's already logged by the Delete function itself as you can see in the code below. I just improved that function by adding the `who` also because it was only logging the option.

```
func (m *MonStore) Delete(who, option string) error {
	logger.Infof("deleting %q %q option from the mon configuration database", who, option)
	args := []string{"config", "rm", who, normalizeKey(option)}
	cephCmd := client.NewCephCommand(m.context, m.clusterInfo, args)
	out, err := cephCmd.RunWithTimeout(exec.CephCommandsTimeout)
	if err != nil {
		return errors.Wrapf(err, "failed to delete ceph config in the centralized mon configuration database. output: %s",
			string(out))
	}

	logger.Infof("successfully deleted %q option from the mon configuration database", option)
	return nil
}

```

Comment (travisn) [reply pkg/operator/ceph/cluster/mgr/dashboard.go:128]: Is that error logged? It looks like if err is returned, it just sets `success = false` and continues. We don't know what the failure was.

Comment (rkachach) [reply pkg/operator/ceph/cluster/mgr/dashboard.go:128]: @travisn the error is `logged` by the Delete function itself:

```
		for _, key := range mgrKeysToDelete {
			err := monStore.Delete(mgrDaemonID, key)
```
and at the end we log whether all the deletes went OK or there was some failure by: 

```
	if success {
		logger.Info("All per-daemon mgr configuration has been deleted successfully.")
	} else {
		logger.Error("At least one delete operation failed while trying to delete per-daemon mgr configuration.")
	}
```

I think from the point of view of logging we should be OK.

Comment (travisn) [reply pkg/operator/ceph/cluster/mgr/dashboard.go:128]: Whenever there is a failure, we need to log the specific `err`, or else we really don't know how to troubleshoot. So if we log the specific err here on this line, we don't need to log the general error on line 136.

Comment (rkachach) [reply pkg/operator/ceph/cluster/mgr/dashboard.go:128]: @travisn this already done on the `Delete` function:

```
	out, err := cephCmd.RunWithTimeout(exec.CephCommandsTimeout)
	if err != nil {
		return errors.Wrapf(err, "failed to delete ceph config in the centralized mon configuration database. output: %s",
			string(out))
	}
```
Not sure if I'm following at this point... but if add another log we will be just duplicating the error message!

Comment (travisn) [reply pkg/operator/ceph/cluster/mgr/dashboard.go:128]: The err is returned, but I don't see where the `err` var is logged. It appears the `err` result is ignored other than setting `success = false`, right?

Comment (rkachach) [reply pkg/operator/ceph/cluster/mgr/dashboard.go:128]: @travisn you are right, not sure why I thought that the error was being logged inside the Delete when reading the code again and again. It's just being returned. I'll fix it.

**Review by travisn — APPROVED** [review]
